queries/petrol.py

```diff
@@ -0,0 +1,297 @@
+       (at your option) any later version.
+       This program is distributed in the hope that it will be useful,
+       but WITHOUT ANY WARRANTY; without even the implied warranty of
+       MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+       GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see http://www.gnu.org/licenses/.
+
+
+    This module handles petrol-related queries.
+
+"""
+
+# TODO: "Hver er ódýrasta bensínstöðin innan X kílómetra?"
+
+
+import logging
+import cachetools
+import random
+
+from geo import distance
+from queries import (
+    query_json_api,
+    format_icelandic_float,
+    gen_answer,
+    distance_desc,
+    krona_desc,
+)
+
+
+_PETROL_QTYPE = "Petrol"
+
+
+TOPIC_LEMMAS = ["bensín", "bensínstöð", "bensínlítri", "dísel"]
+
+
+def help_text(lemma):
+    """ Help text to return when query.py is unable to parse a query but
+        one of the above lemmas is found in it """
+    return "Ég get svarað ef þú spyrð til dæmis: {0}?".format(
+        random.choice(
+            (
+                "Hvar er næsta bensínstöð",
+                "Hvar fæ ég ódýrasta bensínið",
+                "Hvar er ódýrt að fylla tankinn",
+                "Hvar fæ ég ódýrt bensín í nágrenninu",
+            )
+        )
+    )
+
+
+# This module wants to handle parse trees for queries
+HANDLE_TREE = True
+
+# The context-free grammar for the queries recognized by this plug-in module
+GRAMMAR = """
+
+Query →
+    QPetrol
+
+QPetrol → QPetrolQuery '?'?
+
+QPetrolQuery →
+    QPetrolClosestStation | QPetrolCheapestStation | QPetrolClosestCheapestStation
+
+QPetrolClosestStation →
+    "bensín" QPetrolNearMe?
+    | QPetrolStation QPetrolNearMe
+    | "hvar" "fæ" "ég" "bensín" QPetrolNearMe?
+    | "hvar" "get" "ég" "fengið" "bensín" QPetrolNearMe?
+    | "hvar" "get" "ég" "keypt" "bensín" QPetrolNearMe?
+    | "hvar" "get" "ég" "fyllt" "á"? "tankinn" QPetrolNearMe?
+    | "hvar" "er" QPetrolClosest QPetrolStation
+    | "hver" "er" QPetrolClosest QPetrolStation
+    | "hvaða" QPetrolStation "er" QPetrolNearMe
+
+QPetrolCheapestStation →
+    "ódýrasta" "bensínið"
+    | "hvaða" "bensínstöð" "er" "ódýrust"
+    | "hvaða" "bensínstöð" "er" "með" "ódýrasta" "bensínið"
+    | "hvaða" "bensínstöð" "er" "með" "ódýrasta" "bensínlítrann"
+    | "hvaða" "bensínstöð" "er" "ódýrust"
+    | "hvaða" "bensínstöð" "er" "með" "lægsta" "verðið"
+    | "hvaða" "bensínstöð" "er" "með" "lægsta" "verðið" "á" "bensíni"
+    | "hvar" "er" "bensín" "ódýrast"
+    | "hvar" "er" "bensínið" "ódýrast"
+    | "hvar" "er" "bensínlítrinn" "ódýrastur"
+    | "hvar" "fæ" "ég" "ódýrasta" "bensínið"
+    | "hvar" "fæ" "ég" "ódýrasta" "bensínlítrann"
+    | "hvar" "er" "ódýrast" "að" "fylla" "á"? "tankinn"
+
+QPetrolClosestCheapestStation →
+    "ódýrt" "bensín" QPetrolNearMe?
+    | "hvar" "fæ" "ég" "ódýrt" "bensín" QPetrolNearMe?
```

Comment:
Why is QPetrolNearMe optional? If left out, shouldn't this be a QPetrolCheapestStation?

Reply:
A person in Akureyri asking "Hvar fæ ég ódýrt bensín?" is unlikely to want an answer telling him to go to Costco in Reykjavík some 500km away :).